testing/monkey/monkey_dev.c

```diff
@@ -0,0 +1,327 @@
+{
+  bool retval = false;
+  struct touch_sample_s sample;
+
+  int nbytes = read(fd, &sample, sizeof(sample));
+
+  if (nbytes == sizeof(sample))
+    {
+      retval = true;
+      uint8_t touch_flags = sample.point[0].flags;
+
+      if (touch_flags & TOUCH_DOWN || touch_flags & TOUCH_MOVE)
+        {
+          *x = sample.point[0].x;
+          *y = sample.point[0].y;
+          *touch_down = true;
+        }
+      else if (touch_flags & TOUCH_UP)
+        {
+          *x = 0;
+          *y = 0;
+          *touch_down = false;
+        }
+    }
+
+  return retval;
+}
+
+/****************************************************************************
+ * Name: button_read
+ ****************************************************************************/
+
+static bool button_read(int fd, FAR uint32_t *value)
+{
+  btn_buttonset_t buttonset;
+
+  int ret = read(fd, &buttonset, sizeof(buttonset));
+  if (ret < 0)
+    {
+      return false;
+    }
+
+  *value = buttonset;
+  return true;
+}
+
+/****************************************************************************
+ * Name: monkey_dev_create
+ ****************************************************************************/
+
+FAR struct monkey_dev_s *monkey_dev_create(FAR const char *dev_path,
+                                           enum monkey_dev_type_e type)
+{
+  FAR struct monkey_dev_s *dev;
+  int oflag;
+  int fd;
+
+  MONKEY_ASSERT_NULL(dev_path);
+
+  dev = malloc(sizeof(struct monkey_dev_s));
+  MONKEY_ASSERT_NULL(dev);
+  memset(dev, 0, sizeof(struct monkey_dev_s));
+
+  if (MONKEY_IS_UINPUT_TYPE(type))
+    {
+      oflag = O_RDWR | O_NONBLOCK;
+    }
+  else
+    {
+      oflag = O_RDONLY | O_NONBLOCK;
+    }
+
+  fd = open(dev_path, oflag);
+  if (fd < 0)
+    {
+      MONKEY_LOG_ERROR("open %s failed: %d", dev_path, errno);
+      goto failed;
+    }
+
+  dev->type = type;
+  dev->fd = fd;
+
+  MONKEY_LOG_NOTICE("open %s success, fd = %d, type: %s",
+                    dev_path, fd, monkey_dev_type2name(type));
+
+  return dev;
+
+failed:
+  if (fd > 0)
+    {
+      close(fd);
+    }
+
+  if (dev)
+    {
+      free(dev);
+    }
+
+  return NULL;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: monkey_dev_delete
+ ****************************************************************************/
+
+void monkey_dev_delete(FAR struct monkey_dev_s *dev)
+{
+  MONKEY_ASSERT_NULL(dev);
+
+  if (dev->fd > 0)
+    {
+      /* Reset input state */
+
+      struct monkey_dev_state_s state;
+      memset(&state, 0, sizeof(state));
+      state.type = dev->type;
+      monkey_dev_set_state(dev, &state);
+
+      MONKEY_LOG_NOTICE("close fd: %d", dev->fd);
+      close(dev->fd);
+      dev->fd = -1;
```

Comment:
not needed since is followed by `free` unconditionally

Reply:
done